File: Algos & templates/题目/基础/双指针 离散化 差分 前缀和/双指针 离散化 区间合并/LC2763.go
package main

func sumImbalanceNumbers(nums []int) (ans int) {
	n := len(nums)
	right := make([]int, n+5)
	idx := make([]int, n+5)
	for i := 0; i <= n + 1; i++ {
		idx[i] = n + 1
	}
	for i := n; i >= 1; i-- {
		x := nums[i-1]
		right[i] = min(idx[x], idx[x-1])
		idx[x] = i
	}
	for i := range idx {
		idx[i] = 0
	}
	for i := 1; i <= n; i++ {
		x := nums[i-1]
		ans += (i - idx[x-1]) * (right[i] - i)
		idx[x] = i
	}
	return ans - n*(n+1)/2
}

func min(a, b int) int {
	if b < a {
		return b
	}
	return a
}

File: 基础/双指针 离散化 差分 前缀和/双指针 离散化 区间合并/LC2763.py
from sortedcontainers import SortedList
from collections import Counter


class Solution:
    def sumImbalanceNumbers(self, nums: List[int]) -> int:
        n = len(nums);
        ans = 0;
        for i in range(0, n):
            s = SortedList();
            s.add(nums[i]);
            cnt = 0
            for j in range(i + 1, n):
                if (not s.count(nums[j])):
                    f1 = s.count(nums[j] - 1);
                    f2 = s.count(nums[j] + 1);
                    if (f1 and f2):
                        cnt -= 1
                    elif(( not f1) and ( not f2)):
                        cnt += 1
                    s.add(nums[j])
                ans += cnt

        return ans